Add docker-based test scripts

pkolbus opened the pull request:
## Description
Adds a Dockerfile and some test scripts per #2313 

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
I picked Ubuntu 16.04 LTS since it's an older release that I'm familiar with, and one with some support left in it. If you'd be happier with a different base distribution let me know. I'm also not sure if there would be more preferred versions of openssl and gnutls.

This is my first PR, so there's probably something else I'm missing about the contribution process or patch too. Happy to take care of anything you flag in review.

Backporting to 2.16 may be useful but not required.

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [ ] Backported


## Steps to test or reproduce
On a Linux x86_64 box with a recent Docker properly installed, run tests/scripts/all-in-docker.sh or tests/scripts/basic-in-docker.sh.

RonEld commented:
@pkolbus Thank you for your contribution!
According to your description, this is a company contribution.
Your company will need to supply a corporate CLA, signed by someone authorized on behalf of your company. 
You can find the [agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us, and is applicable if this is a corporate contribution.
Thanks again for your contribution!

k-stachowiak commented on tests/docker/Dockerfile:
```diff
@@ -0,0 +1,98 @@
+
+# Build a static, legacy openssl from sources with sslv3 enabled
+# Based on https://gist.github.com/bmaupin/8caca3a1e8c3c5686141 (build-openssl.sh)
+RUN cd /tmp \
```
I'd like to suggest another OpenSSL versioning scheme. Based on our internally used Dockerfile, I suggest approach like:

    RUN wget -q https://www.openssl.org/source/old/1.0.2/openssl-1.0.2g.tar.gz && \
        tar -zxf openssl-1.0.2g.tar.gz && cd openssl-1.0.2g && \
        ./config --openssldir=/usr/local/openssl-1.0.2g && \
        make clean && make && make install && cd .. && \
        rm -rf openssl-1.0.2g*

I don't mind piping the archive straight into `tar` - that is neat, but please note the installation path; by using only the `--openssldir` configuration argument we get the library installed in a directory named with a proper version number. I find this cleaner than just copying the binary with a new name, straight into `/usr/local/bin`.

You can then follow the command above with:

    ENV OPENSSL=/usr/local/openssl-1.0.2g/bin/openssl

...and have an environmental variable ready for use with Mbed TLS test scripts (as they look for it if you don't provide it with the CLI arguments).

It would be great if you could do it for both the OpenSSL versions.

pkolbus replied:
Done. Turns out it's actually `--prefix` that controls where the binary goes (`--openssldir` is for the data), and `make clean` seems not needed given the directory is new. Applied the same for the legacy gnutls.

pkolbus replied:
> Turns out it's actually `--prefix` that controls where the binary goes (`--openssldir` is for the data)

Correction: it depends on the OpenSSL version. `openssldir` is correct for 1.0.2g.

k-stachowiak commented on tests/scripts/all-in-docker.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash -eu
```
This is a really neat idea. Could you also provide such scripts for `tests/ssl-opt.sh` and `tests/compat.sh` ?

pkolbus replied:
Sure thing. Done for both.

k-stachowiak commented on tests/scripts/docker.sh:
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash -eu
```
I think this script could use a better name, like `prepare_docker.sh` or `docker_env.sh`.

pkolbus replied:
Done (docker_env.sh)

pkolbus commented:
Thanks for the feedback. Will address the other comments soon, but one question about:

> did you think about a scheme of supporting different guest systems?

I have other projects that do this, and it wouldn’t be too hard to support. Do you have any thoughts or preferences about how a user would select the guest? Environment variable, command line option? Should there be a default or is it required to pick?

pkolbus commented on tests/docker/Dockerfile:
```diff
@@ -0,0 +1,98 @@
+    && wget https://ftp.gnu.org/gnu/nettle/nettle-2.7.1.tar.gz -qO- | tar xz \
+    && cd nettle-2.7.1 \
+    && ./configure --disable-documentation \
+    && make -j \
```
`all.sh` also applies `-j` unless overridden, and `-j` makes for a significant speedup here. Timings on my PC:

Step | With -j | Without
----- | --- | ---
nettle | 12.4s | 43.7s
gnutls-3.3.18 | 46.9s | 78.7s
openssl-1.1.1a | 36.3s | 164.7s

The CPU usage is somewhat bursty so it _should_ cooperate decently enough with other processes. But I can expose a build arg into docker_env.sh to support overrides. Looking more closely it should probably be `-j $(nproc)` anyway.

As to broken makefiles- I've run these make commands a couple dozen times in the course of developing this PR. The only one that has failed for me is openssl-1.0.2g and I've special-cased it.

That has me thinking though- MAKEFLAGS should propagate into the `docker container run` as well to have the expected effect.

k-stachowiak replied:
Thanks for the additional analysis! Regarding the makeflags - if you mean passing `MAKEFLAGS=-jX`, or even `MAKEFLAGS=-j$(nproc)` to the docker build command, and then override the value for openssl 1.0.2, and whatever doesn't build in parallel, then I think it's a good idea (especially, that in your setup, you can conveniently do this from the bash scripts). However, since, as you have noticed, it may break the builds in some cases, I think it would be the safest path, to have a CLI switch to disable all the `make` parallelization at the scripts level. It should also be configurable (on/off) in case somebody wants to use the plain dockerfile (i.e. it shouldn't be only possible via the bash scripts).

pkolbus replied:
Parallelism of make during the `docker image build`  is now controlled with the `--build-arg MAKEFLAGS_PARALLEL` (in `docker_env.sh`). docker_env will also pass MAKEFLAGS to the `docker container run`, if it is set.

pkolbus commented on tests/docker/Dockerfile:
```diff
@@ -0,0 +1,98 @@
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt update \
```
Done

pkolbus commented on tests/docker/Dockerfile:
```diff
@@ -0,0 +1,98 @@
+RUN cd /tmp \
+    && wget https://openssl.org/source/openssl-1.1.1a.tar.gz -qO- | tar xz \
+    && cd openssl-1.1.1a \
+    && ./config --prefix=`pwd`/local --openssldir=/usr/lib/ssl no-shared \
```
Done- I think. I'm not very familiar with dtags or rpath though, so I'm not sure how to test this bit. Is there a test I can run to confirm it's working as intended? Or is this something that's only in your internal infrastructure?

pkolbus commented:
I've rebased and addressed the review feedback, so this is ready for review again.

Notes:
- I've added an extension point for environment variable MBEDTLS_DOCKER_GUEST, if this is set then a different Dockerfile can be used. Right now the only option is the default (`xenial`).
- I've also updated the Dockerfile to account for conversion to python3 (#2464).

pkolbus commented on tests/compat-in-docker.sh:
```diff
@@ -0,0 +1,50 @@
+    -e GNUTLS_SERV \
+    -e OPENSSL_CMD \
+    -e OSSL_NO_DTLS \
+    tests/compat.sh \
```
The existing `compat.sh` and `ssl-opt.sh` scripts don't do any `make` so I was under the impression that the `*-in-docker.sh` wrappers wouldn't either. Are you suggesting that the `*-in-docker.sh` versions go beyond this -- to be able to build the programs _in the container_ -- to have better assurance of compatibility with the container's ld and libc?

pkolbus replied:
@k-stachowiak, any thoughts? Or maybe what we really need is a `make-in-docker.sh`?

k-stachowiak replied:
Sorry, I forgot to address this question. I'm afraid building in the container is a must, since, as I mentioned, I ran into an actual binary compatibility issues. I think the question would rather be, how to do it.

The fact that we need to work with is, that `ssl-opt.sh` and `compat.sh` actually rely on the client and server examples having been built.

We can express it by putting make inside the docker scripts, but `make-in-docker.sh` looks like a really neat an generic solution.

@Patater This seems to require a design review. What do you think such a battery of "`...-in-docker.sh`" scripts?

Patater replied:
@pkolbus For `make-in-docker.sh`, what is proposed method of saving the generated binaries for re-use in a subsequent `...-in-docker.sh` script? Would the `make-in-docker.sh` script populate the host filesystem with the output? Is that counter-intuitive? Should there be a dedicated Docker volume for the output binaries?

pkolbus replied:
I've added `make-in-docker.sh` with the latest rebase. The docker argument `--volume $PWD:$PWD` (in `docker_env.sh`) causes the worktree to be shared with the container, so any build outputs in the worktree end up in the same place on the host. I've found this mechanism to work well in practice on other projects, and be intuitive (since things like logs and other failure state are easy to find and understand).

k-stachowiak replied:
I concur with that approach. I find it consistent with different kind of builds like `make` vs in tree `cmake`, which by default interfere with each other, but which can be worked around.

The approach suggested in this PR enables creating another script, e.g. `cmake-out-of-tree-in-docker.sh` which will do `mkdir build; cd build; cmake .. $@`, which should prevent overwriting local artifacts with the "in-docker" ones.

The bottom line is, I find this approach flexible and extensible.

pkolbus commented:
> the OpenSSL and GNU TLS versions here are not completely in line with what we use internally
@k-stachowiak - can you identify the versions used internally? It may well be easier to get it right the first time, especially if I've got other review comments to resolve.

k-stachowiak commented:
@pkolbus Sure! These are the versions that we currently use (in the parentheses, I have summarized the preferred auto tools configuration commands if we have any; the ones for gnutls are shared between versions, and we have specific ones for openssl):
```
* GNU TLS (./configure --prefix=/usr/local/gnutls-<version> --exec_prefix=/usr/local/gnutls-<version> --disable-shared --disable-guile)
** GNUTLS: 3.4.10
** GNUTLS_LEGACY: 3.3.8
** GNUTLS_NEXT: 3.6.5
* OpenSSL
** OPENSSL: 1.0.2g
** OPENSSL_LEGACY: 1.0.1j (./config --openssldir=/usr/local/openssl-<version>)
** OPENSSL_NEXT: 1.1.1a or later (./config --prefix=/usr/local/openssl-<version> -Wl,--enable-new-dtags,-rpath,'$(LIBRPATH)')```

pkolbus commented:
@k-stachowiak @Patater I think I've addressed all of your comments and questions; this is ready for re-review.

Patater commented on tests/compat-in-docker.sh:
```diff
@@ -0,0 +1,52 @@
+#
+# This file is part of mbed TLS (https://tls.mbed.org)
+#
+# Copyright (c) 2019, ARM Limited, All Rights Reserved
```
We should use our standard license header here, and in other new files, being sure to mention Apache 2.

See https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/aes.h#L23-L38

pkolbus replied:
Done (I think) in commit 31f0081

pkolbus commented:
This is ready for review again.

pkolbus commented:
Sorry, missed a recently added check for mingw-gcc 6+. Since xenial doesn't have this, `all.sh` just skips the test. Added a patch to switch to bionic so that this runs.

Ready for review again.

pkolbus commented on tests/docker/xenial/Dockerfile:
```diff
@@ -0,0 +1,153 @@
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt update \
```
Done

pkolbus commented:
This should be ready for review again.
